File: togo/internal/cli/run_test.go
package cli

import (
	"strings"
	"testing"
)

func TestSelectMainClass(t *testing.T) {
	// Configured mainClass always wins, even when detection is ambiguous or empty.
	if mc, reason := SelectMainClass([]string{"com.app.A", "com.app.B"}, "com.app.Chosen"); mc != "com.app.Chosen" || reason != "" {
		t.Errorf("configured: got (%q, %q)", mc, reason)
	}
	if mc, reason := SelectMainClass(nil, "com.app.Chosen"); mc != "com.app.Chosen" || reason != "" {
		t.Errorf("configured with no detection: got (%q, %q)", mc, reason)
	}
	// A single detected entry point is used.
	if mc, reason := SelectMainClass([]string{"com.app.Main"}, ""); mc != "com.app.Main" || reason != "" {
		t.Errorf("single: got (%q, %q)", mc, reason)
	}
	// None detected is an error.
	if mc, reason := SelectMainClass(nil, ""); mc != "" || !strings.Contains(reason, "no class declares a main") {
		t.Errorf("none: got (%q, %q)", mc, reason)
	}
	// Ambiguous detection lists the candidates.
	mc, reason := SelectMainClass([]string{"com.app.A", "com.app.B"}, "")
	if mc != "" || !strings.Contains(reason, "com.app.A, com.app.B") {
		t.Errorf("ambiguous: got (%q, %q)", mc, reason)
	}
}

File: src/cli/run.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { selectMainClass } from "./run.ts";

test("selectMainClass: configured mainClass always wins", () => {
  expect(selectMainClass(["com.app.A", "com.app.B"], "com.app.Chosen")).toEqual({
    mainClass: "com.app.Chosen",
  });
  // even when nothing was detected
  expect(selectMainClass([], "com.app.Chosen")).toEqual({ mainClass: "com.app.Chosen" });
});

test("selectMainClass: a single detected entry point is used", () => {
  expect(selectMainClass(["com.app.Main"], undefined)).toEqual({ mainClass: "com.app.Main" });
});

test("selectMainClass: none detected is an error", () => {
  const r = selectMainClass([], undefined);
  expect(r).toEqual({ error: expect.stringMatching(/no class declares a main/) });
});

test("selectMainClass: ambiguous detection lists the candidates", () => {
  const r = selectMainClass(["com.app.A", "com.app.B"], undefined);
  expect(r).toEqual({ error: expect.stringMatching(/com\.app\.A, com\.app\.B/) });
});
